internal/cloud/realm/client.go

```diff
@@ -16,17 +19,37 @@ const (
 
 // Client is a Realm client
 type Client interface {
-	Authenticate(publicAPIKey, privateAPIKey string) (AuthResponse, error)
+	Authenticate(publicAPIKey, privateAPIKey string) (Session, error)
+	GetAuthProfile() (AuthProfile, error)
+	GetAppsForUser() ([]App, error)
+	GetApps(groupID string) ([]App, error)
 	Status() error
 }
 
 // NewClient creates a new Realm client
 func NewClient(baseURL string) Client {
-	return &client{baseURL}
+	return &client{baseURL: baseURL}
+}
+
+// NewAuthClient creates a new Realm client with a session used for Authorization
+func NewAuthClient(baseURL string, session Session) Client {
+	return &client{baseURL, session}
 }
 
 type client struct {
 	baseURL string
+	session Session
+}
+
+func (c *client) doJSON(method, path string, payload interface{}, options api.RequestOptions) (*http.Response, error) {
+	body, err := json.Marshal(payload)
+	if err != nil {
+		return nil, err
+	}
+	options.Body = bytes.NewReader(body)
+	options.ContentType = api.MediaTypeApplicationJSON
+
+	return c.do(method, path, options)
 }
 
 func (c *client) do(method, path string, options api.RequestOptions) (*http.Response, error) {
```

Comment:
do we even need this func? seems like a little bit too much abstraction for this small codebase. I understand we do something similar in baas, but thoughts on removing this piece here? The most important piece in this func to me is the auth, and that's already abstracted into its own func.

Reply:
I imagine we'd be using `c.do` quite a bit for any `GET`/`DELETE` requests, and it handles both setting auth and adding the `X-BAAS-Request-Origin` header.

to manage even just that for every API call we're making seems like a useful abstraction to me.  for what its worth, I see this function eventually expanding to handle some of the response handling as well (the first thing that comes to mind is when an `HttpUnauthorized` comes back, to either automatically attempt a refresh or clear the `profile`'s access token value (to indicate the session is no longer active)

Reply:
I am fine with the current abstraction, but it may be worth revisiting further down the line as we expand the code as Nick mentioned